pkg/controllers/dnsrecord/dnsrecord_controller.go

```diff
@@ -94,6 +102,11 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	reason = "ProviderSuccess"
 	message = "Provider ensured the managed zone"
 
+	DNSProvider, err = r.DNSProv.CreateDNSProvider(ctx, managedZone)
```

Comment:
as mentioned I think I would switch this to be DNSPRovider.Factory() as it is returns a different impl of the provider interface depending on the input